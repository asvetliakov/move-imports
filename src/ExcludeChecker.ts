import fs from "fs";
import ignore from "ignore";
import path from "path";

// because interface is not exported
interface Ignore {
    add(pattern: string | Ignore | string[] | Ignore[]): Ignore;
    filter(paths: string[]): Ignore;
    createFilter(): (path: string) => Ignore;
    ignores(pathname: string): boolean;
}

export class ExcludeChecker {
    /**
     * Globs specified in gitignore, normalized
     */
    private gitIgnoreGlobs: string[] = [];

    /**
     * Workspace root
     */
    private workspaceRoot: string;

    /**
     * Ignore globs specified in configuration
     */
    private confGlobs: string[] = [];

    /**
     * Ignore instance
     */
    private ignore: Ignore;

    /**
     * Creates an instance of ExcludeChecker.
     *
     * @param workspaceRoot
     * @param confGlobs
     * @param readGitIgnore
     */
    public constructor(workspaceRoot: string, confGlobs: string[]) {
        this.workspaceRoot = workspaceRoot;
        this.confGlobs = confGlobs;
        this.ignore = ignore().add(confGlobs);
    }

    /**
     * Read git ignore file
     */
    public async readGitIgnoreExcludes(): Promise<void> {
        try {
            const lines = await new Promise<string>((res, rej) =>
                fs.readFile(path.resolve(this.workspaceRoot, ".gitignore"), "utf8", (e, data) => e ? rej(e) : res(data)),
            );
            for (const line of lines.replace(/\r\n/g, "\n").split("\n")) {
                // Skip empty lines
                if (line.trim() === "") {
                    continue;
                }
                // skip comments
                if (/^\s*#.*/.test(line)) {
                    continue;
                }
                this.gitIgnoreGlobs.push(line.trim());
            }
            if (this.gitIgnoreGlobs.length > 0) {
                this.ignore.add(this.gitIgnoreGlobs);
            }
        } catch {
            // ignore for now
        }
    }

    /**
     * Return full list of ignore globs
     *
     * @param inverted
     */
    public getIgnoreGlobs(inverted: boolean = false): string[] {
        const ignores = [...this.confGlobs, ...this.gitIgnoreGlobs];
        return inverted ? ignores.map(i => i.startsWith("!") ? i.substr(1) : "!" + i) : ignores;
    }

    /**
     * Return true if given file path is excluded either by gitignore or configuration ignore
     *
     * @param filePath
     * @returns
     */
    public isFileExcluded(filePath: string): boolean {
        // convert to relative path -> /a/b/c.js -> b/c.js if workspace root is /a, need for ignore
        if (filePath.startsWith(this.workspaceRoot)) {
            filePath = path.relative(this.workspaceRoot, filePath);
        }
        return this.ignore.ignores(filePath);
    }
}
