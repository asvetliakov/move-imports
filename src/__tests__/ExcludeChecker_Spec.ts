import fsMock from "mock-fs";
import { ExcludeChecker } from "../ExcludeChecker";

let checker: ExcludeChecker;
beforeEach(() => {
    checker = new ExcludeChecker("/u/workspace", ["node_modules", "jspm_modules", "a", "b/*", "!b/c.js"]);
});

afterEach(() => {
    fsMock.restore();
});

it("Checks if the file is excluded", () => {
    expect(checker.isFileExcluded("/u/workspace/a.js")).toBeFalsy();
    expect(checker.isFileExcluded("a.js")).toBeFalsy();

    expect(checker.isFileExcluded("/u/workspace/b.js")).toBeFalsy();
    expect(checker.isFileExcluded("b.js")).toBeFalsy();

    expect(checker.isFileExcluded("/u/workspace/node_modules/a/index.js")).toBeTruthy();
    expect(checker.isFileExcluded("node_modules/a/index.js")).toBeTruthy();

    expect(checker.isFileExcluded("/u/workspace/jspm_modules/a.js")).toBeTruthy();
    expect(checker.isFileExcluded("jspm_modules/a.js")).toBeTruthy();

    expect(checker.isFileExcluded("/u/workspace/a/a.js")).toBeTruthy();
    expect(checker.isFileExcluded("a/a.js")).toBeTruthy();

    expect(checker.isFileExcluded("/u/workspace/b/a.js")).toBeTruthy();
    expect(checker.isFileExcluded("b/a.js")).toBeTruthy();

    expect(checker.isFileExcluded("/u/workspace/b/c.js")).toBeFalsy();
    expect(checker.isFileExcluded("b/c.js")).toBeFalsy();
});

it("Checks if the file is excluded by using git ignore file", async () => {
    fsMock({
        "/u/workspace/.gitignore": `
        # Comment
        node_modules
        **/c
        **/index.*
        # Comment
        e/*
        !e/e.js
        `,
    });
    await checker.readGitIgnoreExcludes();
    expect(checker.isFileExcluded("/u/workspace/node_modules")).toBeTruthy();
    expect(checker.isFileExcluded("node_modules")).toBeTruthy();

    expect(checker.isFileExcluded("/u/workspace/index.js")).toBeTruthy();
    expect(checker.isFileExcluded("index.js")).toBeTruthy();

    expect(checker.isFileExcluded("/u/workspace/a/b/w/index.js")).toBeTruthy();
    expect(checker.isFileExcluded("a/b/w/index.js")).toBeTruthy();

    expect(checker.isFileExcluded("/u/workspace/a/b/c/a.js")).toBeTruthy();
    expect(checker.isFileExcluded("a/b/c/a.js")).toBeTruthy();

    expect(checker.isFileExcluded("/u/workspace/e/a.js")).toBeTruthy();
    expect(checker.isFileExcluded("e/a.js")).toBeTruthy();

    expect(checker.isFileExcluded("/u/workspace/e/e.js")).toBeFalsy();
    expect(checker.isFileExcluded("e/e.js")).toBeFalsy();

    expect(checker.isFileExcluded("/u/workspace/y.js")).toBeFalsy();
    expect(checker.isFileExcluded("y.js")).toBeFalsy();
});

it("Returns globs", async () => {
    fsMock({
        "/u/workspace/.gitignore": `
        # Comment
        node_modules
        **/c
        **/index.*
        # Comment
        e/*
        !e/e.js
        `,
    });
    await checker.readGitIgnoreExcludes();
    expect(checker.getIgnoreGlobs()).toMatchSnapshot();
    expect(checker.getIgnoreGlobs(true)).toMatchSnapshot();
});
