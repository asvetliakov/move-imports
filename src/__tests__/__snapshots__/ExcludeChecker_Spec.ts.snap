// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Returns globs 1`] = `
Array [
  "node_modules",
  "jspm_modules",
  "a",
  "b/*",
  "!b/c.js",
  "node_modules",
  "**/c",
  "**/index.*",
  "e/*",
  "!e/e.js",
]
`;

exports[`Returns globs 2`] = `
Array [
  "!node_modules",
  "!jspm_modules",
  "!a",
  "!b/*",
  "b/c.js",
  "!node_modules",
  "!**/c",
  "!**/index.*",
  "!e/*",
  "e/e.js",
]
`;
